module/os/linux/zfs/zvol_os.c

```diff
@@ -54,7 +54,7 @@ static unsigned int zvol_prefetch_bytes = (128 * 1024);
 static unsigned long zvol_max_discard_blocks = 16384;
 
 #ifndef HAVE_BLKDEV_GET_ERESTARTSYS
-static const unsigned int zvol_open_timeout_ms = 1000;
+static unsigned int zvol_open_timeout_ms = 1000;
```

Comment:
Should we consider increasing the default value as well?  How bad can it be in practice?

Reply:
Maybe, we have a workload hitting it, but I haven't tried tuning to see how high we need it. I'm working on that now..

Do you have suggestions on a value I should try first?

I need to double check, but I think maybe there was no limit, when the retry was in the kernel? I'll read the old kernel code in an hour..

Reply:
yea, so we went from a 5.4 based kernel.. to a 5.15 based kernel.. from what I can tell, the 5.4 kernel logic was this:
```
1557  restart:
1558
1559         ret = -ENXIO;
1560         disk = bdev_get_gendisk(bdev, &partno);
1561         if (!disk)
1562                 goto out;
1563
1564         disk_block_events(disk);
1565         mutex_lock_nested(&bdev->bd_mutex, for_part);
1566         if (!bdev->bd_openers) {
1567                 first_open = true;
1568                 bdev->bd_disk = disk;
1569                 bdev->bd_queue = disk->queue;
1570                 bdev->bd_contains = bdev;
1571                 bdev->bd_partno = partno;
1572
1573                 if (!partno) {
1574                         ret = -ENXIO;
1575                         bdev->bd_part = disk_get_part(disk, partno);
1576                         if (!bdev->bd_part)
1577                                 goto out_clear;
1578
1579                         ret = 0;
1580                         if (disk->fops->open) {
1581                                 ret = disk->fops->open(bdev, mode);
1582                                 if (ret == -ERESTARTSYS) {
1583                                         /* Lost a race with 'disk' being
1584                                          * deleted, try again.
1585                                          * See md.c
1586                                          */
1587                                         disk_put_part(bdev->bd_part);
1588                                         bdev->bd_part = NULL;
1589                                         bdev->bd_disk = NULL;
1590                                         bdev->bd_queue = NULL;
1591                                         mutex_unlock(&bdev->bd_mutex);
1592                                         disk_unblock_events(disk);
1593                                         put_disk_and_module(disk);
1594                                         goto restart;
1595                                 }
1596                         }
```
I don't see any timeout here.

I think what I'm going to do, is try tuning our reproducing workload up to a 60s timeout, and let that run over the long weekend.. hopefully with it set that high, I won't see any problems..

For this PR, though.. I think I'd just suggest keeping it as-is, and just make it able to be tuned.. this will have no functional impact (i.e. no risk), but make my life easier as I try different values.. and then I can look into fine-tuning the value, and see if I can come up with a recommendation for a better value to use upstream..

Does that sound reasonable?

Reply:
I could probably add some instrumentation, to get the latency of `zvol_open` calls, once I have a timeout large enough to not throw the error.. that might be useful, and help us determine the "perfect" timeout to use as a default.. at least it will be somewhat data-driven that way :p

Reply:
I'm still running the workload with a 60s timeout, but here's some data from the last couple hours:
```
$ sudo bpftrace -e 'kprobe:zvol_open { @start[tid] = nsecs; } kretprobe:zvol_open /@start[tid]/ { @ms = hist((nsecs - @start[tid])/1000); delete(@start[tid]); }'
Attaching 2 probes...
^C

@ms:
[0]                  319 |@@                                                  |
[1]                 1346 |@@@@@@@@@                                           |
[2, 4)              7196 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[4, 8)              1397 |@@@@@@@@@@                                          |
[8, 16)               37 |                                                    |
[16, 32)              19 |                                                    |
[32, 64)              20 |                                                    |
[64, 128)              4 |                                                    |
[128, 256)             6 |                                                    |
[256, 512)             6 |                                                    |
[512, 1K)              4 |                                                    |
[1K, 2K)            1376 |@@@@@@@@@                                           |
[2K, 4K)             960 |@@@@@@                                              |
[4K, 8K)             276 |@                                                   |
[8K, 16K)            211 |@                                                   |
[16K, 32K)           135 |                                                    |
[32K, 64K)            77 |                                                    |
[64K, 128K)           40 |                                                    |
[128K, 256K)          12 |                                                    |
[256K, 512K)          18 |                                                    |
[512K, 1M)            15 |                                                    |
[1M, 2M)              19 |                                                    |
[2M, 4M)              28 |                                                    |
[4M, 8M)              10 |                                                    |
```

Reply:
I was able to tune the timeout to 60s yesterday, and haven't seen any failure since I did that. I'm not sure if we want to use 60s for the default, but just one data point.

Also, here's some more data about the `zvol_open` latency:
```
$ sudo bpftrace -e 'kprobe:zvol_open { @start[tid] = nsecs; } kretprobe:zvol_open /@start[tid]/ { @ms = hist((nsecs - @start[tid])/1000); delete(@start[tid]); }'
Attaching 2 probes...
^C

@ms:
[0]                  453 |@@                                                  |
[1]                 1911 |@@@@@@@@@                                           |
[2, 4)             10667 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[4, 8)              1616 |@@@@@@@                                             |
[8, 16)               58 |                                                    |
[16, 32)              24 |                                                    |
[32, 64)              12 |                                                    |
[64, 128)             13 |                                                    |
[128, 256)             4 |                                                    |
[256, 512)             3 |                                                    |
[512, 1K)              6 |                                                    |
[1K, 2K)            2454 |@@@@@@@@@@@                                         |
[2K, 4K)            1136 |@@@@@                                               |
[4K, 8K)             298 |@                                                   |
[8K, 16K)            158 |                                                    |
[16K, 32K)           114 |                                                    |
[32K, 64K)            55 |                                                    |
[64K, 128K)           29 |                                                    |
[128K, 256K)          24 |                                                    |
[256K, 512K)          23 |                                                    |
[512K, 1M)            21 |                                                    |
[1M, 2M)               1 |                                                    |


$ sudo bpftrace -e 'kprobe:zvol_open { @start[tid] = nsecs; } kretprobe:zvol_open /@start[tid]/ { @ms = hist((nsecs - @start[tid])/1000); delete(@start[tid]); }'
Attaching 2 probes...
^C

@ms:
[0]                 1286 |@@                                                  |
[1]                 6324 |@@@@@@@@@@@                                         |
[2, 4)             29044 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[4, 8)              4699 |@@@@@@@@                                            |
[8, 16)              139 |                                                    |
[16, 32)              51 |                                                    |
[32, 64)              43 |                                                    |
[64, 128)             33 |                                                    |
[128, 256)            23 |                                                    |
[256, 512)            20 |                                                    |
[512, 1K)             16 |                                                    |
[1K, 2K)            7023 |@@@@@@@@@@@@                                        |
[2K, 4K)            3006 |@@@@@                                               |
[4K, 8K)             723 |@                                                   |
[8K, 16K)            531 |                                                    |
[16K, 32K)           278 |                                                    |
[32K, 64K)           153 |                                                    |
[64K, 128K)           96 |                                                    |
[128K, 256K)          40 |                                                    |
[256K, 512K)          18 |                                                    |
[512K, 1M)            15 |                                                    |
[1M, 2M)               8 |                                                    |
```

so maybe I can bump it down to 16s and still avoid the errors.. unfortunately, it doesn't seem easy to try and limit these to only the cases that hit the retry loop..

Reply:
I'll go ahead and merge this change with the existing default.  As you said, at least this will make it easier to test alternate defaults.

Reply:
Thanks @behlendorf